envoy/common/execution_context.h

```diff
@@ -15,13 +16,42 @@ namespace Envoy {
 static constexpr absl::string_view kConnectionExecutionContextFilterStateName =
     "envoy.network.connection_execution_context";
 
+namespace Http {
+struct FilterContext;
+}
+
+namespace Tracing {
+class Span;
+}
+
 class ScopedExecutionContext;
 
 // ExecutionContext can be inherited by subclasses to represent arbitrary information associated
 // with the execution of a piece of code. activate/deactivate are called when the said execution
 // starts/ends. For an example usage, please see
 // https://github.com/envoyproxy/envoy/issues/32012.
 class ExecutionContext : public StreamInfo::FilterState::Object, NonCopyable {
+public:
+  static void setEnabled(bool value) { enabled().store(value, std::memory_order_relaxed); }
+
+  static bool isEnabled() { return enabled().load(std::memory_order_relaxed); }
+
+  static ExecutionContext* fromStreamInfo(OptRef<const StreamInfo::StreamInfo> info) {
+    if (!isEnabled() || !info.has_value()) {
+      return nullptr;
+    }
+    const auto* const_context = info->filterState().getDataReadOnly<ExecutionContext>(
+        kConnectionExecutionContextFilterStateName);
+    return const_cast<ExecutionContext*>(const_context);
+  }
+
+  // Called when enters a scope in which |span| is active.
+  // Returns an object that can do some cleanup when exits the scope.
+  virtual Envoy::Cleanup onScopeEnter(Envoy::Tracing::Span& span) PURE;
+  // Called when enters a scope in which |filter_context| is active.
+  // Returns an object that can do some cleanup when exits the scope.
+  virtual Envoy::Cleanup onScopeEnter(const Http::FilterContext& filter_context) PURE;
```

Comment:
Will make this a common scope flag like `onScopeEnter(absl::string_view type, absl::string_view name) PURE;` be better?

`FilterContext` will have no other info except the filter name in the visible future.

Reply:
This definitely works for filter, for which I only need `FilterContext.config_name`, but this doesn't work for `Tracing::Span` because I need to cast it into `MyTracingSpan`.

So I think our options are:

1.  Use a generic `onScopeEnter(string, string)` plus a special `onScopeEnter(Envoy::Tracing::Span& span)`.  The `ENVOY_EXECUTION_SCOPE` macro needs to be changed to accept both signatures as well.
2. Or, leave things as-is.  If we want to remove `FilterContext.config_name` or `FilterContext` at some point, we'll have to define a new type for `onScopeEnter(filter)` to continue to work.

What do you think?

Reply:
Got it. Then let's leave it as it's.